packages/teleport/src/Discover/Shared/SetupAccess/useUserTraits.ts

```diff
@@ -0,0 +1,314 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { useState, useEffect } from 'react';
+import useAttempt from 'shared/hooks/useAttemptNext';
+
+import { arrayStrDiff } from 'teleport/lib/util';
+import useTeleport from 'teleport/useTeleport';
+import { Option } from 'teleport/Discover/Shared/SelectCreatable';
+
+import { ResourceKind } from '../ResourceKind';
+
+import type { User, UserTraits } from 'teleport/services/user';
+import type { DbMeta, KubeMeta, NodeMeta } from 'teleport/Discover/useDiscover';
+import type { AgentStepProps } from '../../types';
+
+// useUserTraits handles:
+//  - retrieving the latest user (for the dynamic traits) from the backend
+//  - extracting traits into static traits (role-defined) and dynamic traits (user-defined)
+//  - updating user in the backend with the latest dynamic traits
+//  - updating the dynamic traits for our in-memory resource meta object
+//  - provides utility function that makes data objects (type Option) for react-select component
+export function useUserTraits(props: AgentStepProps) {
+  const ctx = useTeleport();
+
+  const [user, setUser] = useState<User>();
+  const { attempt, run, setAttempt, handleError } = useAttempt('processing');
+
+  const isSsoUser = ctx.storeUser.state.authType === 'sso';
+  const canEditUser = ctx.storeUser.getUserAccess().edit;
+  const dynamicTraits = initUserTraits(user);
+
+  // Filter out static traits from the resource that we
+  // queried in a prior step where we discovered the newly connected resource.
+  // The resource itself contains traits that define both
+  // dynamic (user-defined) and static (role-defined) traits.
+  let meta = props.agentMeta;
+  let staticTraits = initUserTraits();
+  switch (props.selectedResourceKind) {
+    case ResourceKind.Kubernetes:
+      const kube = (meta as KubeMeta).kube;
+      staticTraits.kubeUsers = arrayStrDiff(
+        kube.users,
+        dynamicTraits.kubeUsers
+      );
+      staticTraits.kubeGroups = arrayStrDiff(
+        kube.groups,
+        dynamicTraits.kubeGroups
+      );
+      break;
+
+    case ResourceKind.Server:
+      const node = (meta as NodeMeta).node;
+      staticTraits.logins = arrayStrDiff(node.sshLogins, dynamicTraits.logins);
+      break;
+
+    case ResourceKind.Database:
+      const db = (meta as DbMeta).db;
+      staticTraits.databaseUsers = arrayStrDiff(
+        db.users,
+        dynamicTraits.databaseUsers
+      );
+      staticTraits.databaseNames = arrayStrDiff(
+        db.names,
+        dynamicTraits.databaseNames
+      );
+      break;
+
+    default:
+      throw new Error(
+        `useUserTraits.ts:statiTraits: resource kind ${props.selectedResourceKind} is not handled`
+      );
+  }
+
+  useEffect(() => {
+    fetchUserTraits();
+  }, []);
```

Comment:
If leaving the dependencies blank is intentional can you add a comment and a `//eslint ignore` label here so a future dev knows this was an intentional decision.

Reply:
i used this technique: https://gravitational.slack.com/archives/C02DQ1C2BMW/p1667381168976679?thread_ts=1667257043.804519&cid=C02DQ1C2BMW